### DISCUSSION TITLE
Supply alternate server for certain dependencies?

### DISCUSSION THREAD

**Discussion by ifrisbie** [microsoft/vcpkg#24195] [General]
I'm trying to control my environment using versioning but what I'm finding is that since the source server references are provided in the port file - if that server is no longer valid, it won't download.  You end up needing a more up to date baseline reference for when that was fixed.  So if you are attempting to keep an older baseline there seems to be no way around updating it (which I don't want to do - as I think its going to break my ability to build older things).  Is there a way to supply an alternate server so downloads won't fail?  My example is the pcre lib where in Nov 2021 it was updated to point to Sourceforge.  Is there some configuration option where you can identify backup servers to try for download issues?  Or ideally a way to specify it on a per-dependency basis?  If not either of these things is there a way to patch the version captures (buildtree/versioning/...) post grab and fix the problem as part of the build process?

Thanks in advance!

**Comment by dg0yt** [comment]
You can manage caching of downloads via "asset caching" or via manual maintenance of the `downloads` folder.

**Reply by ifrisbie** [reply]
Thanks, I'll look into that.

**Reply by ifrisbie** [reply]
Looks to me like asset caching assumes you already have a viable location for the source and it subsequently went away.  In order to use this I'd need to already have viable binaries to cache - but I'm building a fresh vcpkg.  I would need to manually create the binary cache would I not?  Was hoping there would be an alternate method for identifying the location of actual source.  Unless I'm not understanding the feature.  Regarding "manual maintenance" - I'm looking for something I can script - really don't want to introduce any manual steps in the process.  

 My bigger issue though is that I can't seem to teach the versioning (manifest, dependency specifiers) to download the version of the library I actually need.  Take pcre.  The very latest is 8.45#2 - but if I use a version>= and specify that version, I get an override required notice.  When I add that override and specify that specific version I get the message:

```
-- Running vcpkg install
Error: Cannot resolve a minimum constraint for dependency vcpkg-cmake from pcre:x64-osx.
The dependency was not found in the baseline, indicating that the package did not exist at that time. This may be fixed by providing an explicit override version via the "overrides" field or by updating the baseline.
See `vcpkg help versioning` for more information.
```

As if I never specified it:

```
{
    "name": “myproj”,
    "version": "1.0.0",
    "dependencies": [
        { "name" : "pcre", "version>=": "8.45#2" }
    ],
    "overrides": [
        { "name": "pcre", "version": "8.45#2" }
    ],
    "builtin-baseline": "68a74950d0400f5a803026d0860f49853984bf11"
}

```

The fix I need is in this version - I just can't teach it to take it.  If I omit the override, it gives me a warning that it can't compare versions because the 8.45 ones are "version" encoded where 8.44 are "version-string" encoded (encoded my paraphrasing).  If I could get it to just take that version I wouldn't need to manually build and ultimately cache the one I need.

My understanding of how this is supposed to work is that I shouldn't need to modify my baseline.  The error message said either do override OR modify the baseline.  Well, the example above shows the former - but it doesn't work.  Any help would be appreciated.
